vm/rust/src/juno_state.rs

```
@@ -164,6 +192,83 @@ impl StateReader for JunoStateReader {
+    }
+
+    /// Allocates the given address to the given class hash.
+    /// Raises an exception if the address is already assigned;
+    /// meaning: this is a write once action.
+    fn set_class_hash_at(
+        &mut self,
+        contract_address: ContractAddress,
+        class_hash: ClassHash,
+        ) -> StateResult<()> {
+        let addr = felt_to_byte_array(contract_address.0.key());
+        let class_hash = felt_to_byte_array(&class_hash.0);
+        state_read_err(unsafe { JunoStateSetClassHashAt(self.handle, addr.as_ptr(), class_hash.as_ptr()) })
+    }
+
+    /// Sets the given contract class under the given class hash.
+    fn set_contract_class(
+        &mut self,
+        class_hash: &ClassHash,
+        _contract_class: ContractClass,
+        ) -> StateResult<()> {
+        let class_hash = felt_to_byte_array(&class_hash.0);
+        state_read_err(unsafe { JunoStateSetContractClass(self.handle, class_hash.as_ptr()) })
+    }
+
+    /// Sets the given compiled class hash under the given class hash.
+    fn set_compiled_class_hash(
+        &mut self,
+        class_hash: ClassHash,
+        compiled_class_hash: CompiledClassHash,
+        ) -> StateResult<()> {
+        let class_hash_bytes = felt_to_byte_array(&class_hash.0);
+        let compiled_class_hash_bytes = felt_to_byte_array(&compiled_class_hash.0);
+        state_read_err(unsafe { JunoStateSetCompiledClassHash(self.handle, class_hash_bytes.as_ptr(), compiled_class_hash_bytes.as_ptr()) })
+    }
+
+    fn to_state_diff(&mut self) -> CommitmentStateDiff {
+        // return an empty state diff for now. Not needed when we are directly storing all writes anyway.
+        CommitmentStateDiff {
+            address_to_class_hash: IndexMap::default(),
+            storage_updates: IndexMap::default(),
+            class_hash_to_compiled_class_hash: IndexMap::default(),
+            address_to_nonce: IndexMap::default(),
+        }
+    }
```

Instead of silently returning incomplete data, let's use `unimplemented` here. I don't think it will be called at all.

---

It's called for traces. Do we want to break those handlers in the sequencer?

---

We call it on the `txn_state`, not the JunoState. Traces should still work.